### ISSUE TITLE
[Code health] clang-tidy cleanup

### ISSUE THREAD

**Issue by marcalff** [open-telemetry/opentelemetry-cpp#2053]
Run [clang-tidy](https://clang.llvm.org/extra/clang-tidy/) on the code base, fix issues reported.

**Comment by esigo** [comment]
Once cleaned, it's good to have a CI for this in another issue.

**Comment by jensecrest** [comment]
Hello, I would like to work on this issue!

**Comment by dbarker** [comment]
Hi @jensecrest, Thanks! 

I've logged a few issues: #3977, #3978, #3979, #3980 that would be a good place to start. Please take a look and comment on one you'd like to take on.

**Comment by dbarker** [comment]
The Github clang-tidy action is uploading warning reports and is a good place to look for current clang-tidy warnings on main. 
https://github.com/open-telemetry/opentelemetry-cpp/actions/workflows/clang-tidy.yaml?query=branch%3Amain++

**Comment by thc1006** [comment]
Big thanks to @marcalff and @lalitb for the quick turnarounds on
this thread over the last few days — the cleanup moved fast
because of your reviews.

Posting a status update + handover map.

**State**: `misc-use-internal-linkage` pool now at 98 abiv2 sites
across 63 files, down from ~135 at the start of these PRs (#4115
#4116 #4121 #4122 #4124 #4128 #4129). Distribution:

| Sites per file | File count |
|---|---|
| 4 | 2 |
| 3 | 5 |
| 2 | 19 |
| 1 | 37 |

**Wrap pattern, for anyone picking up the rest:**

```cpp
namespace
{
// existing class/struct/enum declarations stay here
}  // namespace
```

Things I tripped on (sharing so others don't):

1. **clang-tidy v22 flags enums too** — not just classes/structs.
   I missed `enum test_mode` in `func_http_main.cc` (1 site left
   after #4128 wraps the classes). Always grep `^enum` alongside
   `^class` and `^struct`.

2. **Wrap-boundary trap** for files with forward-declared static
   helpers (`static int foo();` early + `static int foo() {...}`
   after `main()`): the wrap MUST exclude the forward declarations
   OR include their definitions. Wrapping declarations alone
   creates `(anonymous namespace)::foo` symbols that fail to link
   against the file-scope `static foo` definitions. Hit this in
   pre-verify on #4128 (12+ undefined references); fixed by
   shrinking wrap to end before the forward decls.

3. **ABI-gated classes need asymmetric ratchet** (-N abiv1 / -M
   abiv2 where M > N). PR #4124 was the first such drop in this
   campaign; it's a defensible pattern if the PR body explains the
   `#if abiv2` gate clearly.

**Top files (4 and 3 sites)** from the post-#4129 abiv2 CI report:

| Sites | File |
|---|---|
| 4 | `api/test/nostd/shared_ptr_test.cc` |
| 4 | `api/test/logs/logger_test.cc` |
| 3 | `sdk/test/trace/tracer_test.cc` |
| 3 | `sdk/test/trace/batch_span_processor_test.cc` |
| 3 | `sdk/test/metrics/async_metric_storage_test.cc` |
| 3 | `sdk/test/logs/batch_log_record_processor_test.cc` |
| 3 | `ext/test/http/url_parser_test.cc` |

The 19 two-site and 37 one-site files are scattered across
`sdk/test/metrics`, `api/test`, `exporters/otlp/test`, and
`exporters/{zipkin,prometheus,elasticsearch}/test`.

For the full per-file list:

```bash
gh run download <run-id> \
  --repo open-telemetry/opentelemetry-cpp \
  --name Report-clang-tidy-all-options-abiv2-preview
```

(Find the run ID via `gh run list --workflow clang-tidy.yaml --branch main --limit 1`.)

**Open to anyone picking these up:** the wrap is mechanical
(~5 lines source + ratchet update + CHANGELOG bullet). Happy to
answer questions if it helps.

I'll be focusing on other areas of the codebase next, but I'll
keep an eye on this thread.

(If a different style of closeout would be more useful — bundle
PR, wiki page, etc. — happy to switch approaches.)

**Comment by itu-itis24-iyigun24** [comment]
I'd like to work on the remaining `bugprone-exception-escape` warnings in `sdk/src/metrics/aggregation/lastvalue_aggregation.cc`, following the non-throwing `get_if` approach used in #4109.
